ouroboros-network/src/Ouroboros/Network/Diffusion.hs

```diff
@@ -310,14 +337,18 @@ runDataDiffusion tracers
                Left sd -> do
                    a <- Socket.getSocketName sd
                    case a of
-                        (Socket.SockAddrUnix path) ->
+                        (Socket.SockAddrUnix path) -> do
+                          traceWith dtDiffusionInitializationTracer $ CreatingSystemdSocketForUnixPath path
                           return (LocalSocket sd, Snocket.localSnocket iocp path)
-                        _unsupportedAddr ->
+                        unsupportedAddr -> do
+                          traceWith dtDiffusionInitializationTracer $ UnsupportedLocalSystemdSocket unsupportedAddr
                           throwIO UnsupportedLocalSocketType
 #endif
                Right addr -> do
                    let sn = Snocket.localSnocket iocp addr
+                   traceWith dtDiffusionInitializationTracer $ CreateSystemdSocketForSnocketPath addr
```

Comment:
think you are more interested to see a log message when `Snocket.open` fails, than searching for `Create` / `Created` messages.  If `Snocket.open` fails and throws an exception, `runLocalServer` will fail, and then `runDataDiffusion` will fail as well.  As a result the node will stop with an exception.  Isn't it enough? If you think so too, then I suggest we remove `CreateSystemdSocketForSnocketPath`.

Reply:
This is mostly because I don't trust that node will actually trust that stop with an exception under all circumstances.

Reply:
Then wrap `runDataDiffusion` top level action with:
```
handle (\(e :: SomeExcpetion) -> withTracer dtDiffusionInitializationTracer (DiffusionErrored e) >> throwIO e $ do ... -- the runDataDiffusion logic follows.
```
You'll see not only that the diffusion broke, but also why.

You could test what node does, just before it calls `runDataDiffusion` you could call `throwIO (userError "HALT")`. For a longer term, we could use `io-sim-classes` package in the `node` package.  This would allow to testing node using `IOSim` and catch errors which you are hunting.

Reply:
The reason I hesitate is I don't know how all the code fits together.

For example:

* are threads forked somewhere, so the exception doesn't propagate to the top-level
* do exceptions get caught somewhere between here and there and don't get rethrown
* or maybe one of the socket functions block and never return
* or maybe there's an error that's not an exception, but it's ignored, so logging on an exception doesn't help

Not knowing the code well, I can't make these judgement calls.  I would be happy to remove any excess logging if they prove to be unfruitful when the failure occurs and the logging doesn't help.

But I your suggestion to add a top-level wrapper seems useful regardless.

Reply:
The sequence seems this: `main` calls `Cardano.Node.Run.runNode` which wraps `Ouroboros.Consensus.Node.run` in `withShutdownHandler`.  The latter call is using `race_` between the main action (`Ouroboros.Consensus.Node.run`) and some monitoring thread.  `Ouroboros.Consensus.Node.run` tail calls `Ouroboros.Network.Diffusion.runDataDiffusion`. All of them are done without using fork, threads are spawned by `Ouroboros.Consensus.Node.run` (for some consensus functionality) and `runDataDiffusion`.  The `main` installs `topLevelExceptionHandler`which seems fine.

`runDataDiffusion` is using `withAsync` to fork all threads and awaits on all asyncs using `waitAnyCancel` (using `waitAny` would be fine too).  It seems to me that exceptions are handled correctly up to `main`.